### ISSUE TITLE
VSS: access denied for "System Volume Information"

### ISSUE THREAD

**Issue by JackHewson** [duplicati/duplicati#3015]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**:  2.0.2.17_canary_2018-01-23
- **Operating system**: Windows 10
- **Backend**: FTP

## Description
I have activated VSS to backup opened files too. My backup source is the whole drive D:\ .
The system folder "System Volume Information" should be ignored.

## Steps to reproduce

- set "snapshot-policy" to "Required".
- Two test iterations - the result is the same!
    - set Default filter to: none (no checkbox is activated) 
   - set Default filter to: all (all checkboxes are activated)
- set backup source Volume Drive D:\ 

- **Actual result**:
Error message:
Failed to process path: D:\System Volume Information\1{xxx8876b-c176-4xxx-b7ae-04046e6ccxxx} => (5) Zugriff verweigert: [\\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy2\System Volume Information\1{xxx8876b-c176-4xxx-b7ae-04046e6ccxxx}],
- **Expected result**:
Directory "System Volume Information" is automatically excluded, as mentioned in https://github.com/duplicati/duplicati/search?utf8=%E2%9C%93&q=%22System+Volume+Information%22&type=

## Workaround
Set own filter "Exclude directories whose name contains" to "System Volume Information", then it works.

**Comment by tygill** [comment]
I wonder if using VSS causes the path that gets checked against the filter to use the VSS path instead of the original path (e.g., the "\?\GLOBALROOT\DEVICE\HarddiskVolumeShadowCopy2\System Volume Information\..." path in the error message)? I'm not familiar enough with where the path might be converted to its VSS form to know for sure. Currently, the "System Volume Information" filter assumes this folder will be on the root of a drive, and so it wouldn't match against that alternate path.

However, I haven't been able to think of a reason why it should need to only match those folders at the root. And I've been wondering for a while if I should change the Windows filters to not care if things are at the root (so they're more useful when drives are mounted on non-Windows machines where they wouldn't have the same drive prefix). So this was enough to push me to do that - I've made a change to make all the Windows default filters not be rooted to a drive, so they should match anything that has the corresponding names (basically, doing your workaround directly).

**Comment by kenkendk** [comment]
I am slightly worried about the non-rooting. It could cause unexpected files to be excluded from the backup. The snapshots report the non-mapped path (i.e. what it would be, had there been no snapshot) and internally map all requests to the corresponding snapshot paths.

I don't think the filters are applied to the internal paths. Have you debugged it and seen why the filters fail to exclude the path?

**Comment by tygill** [comment]
When running from the command line, I'm not able to reproduce the error (System Volume Information, etc. seem to be excluded):

```
C:\Program Files\Duplicati 2>Duplicati.CommandLine.exe backup C:\DuplicatiTest\DefaultFilters.Test.Backup C:\ --default-filters=all --snapshot-policy=required --no-encryption --log-file=C:\DuplicatiTest\DefaultFilters.Test.Backup.log --log-level=information --verbose
Input command: backup
Input arguments:
        C:\DuplicatiTest\DefaultFilters.Test.Backup
        C:\

Input options:
snapshot-policy: required
no-encryption:
log-file: C:\DuplicatiTest\DefaultFilters.Test.Backup.log
log-level: information
verbose:

Backup started at 2/8/2018 2:12:29 PM
Checking remote backup ...
  Listing remote folder ...
Scanning local files ...
  0 files need to be examined (0 bytes)
Including source path: C:\
Adding directory C:\
Excluding path due to filter: C:\$Recycle.Bin\ => <TRIMMED DEFAULT FILTERS>
Excluding path due to filter: C:\Config.Msi\ => <TRIMMED DEFAULT FILTERS>
Storing symlink: C:\Documents and Settings\
Excluding path due to filter: C:\Program Files\ => <TRIMMED DEFAULT FILTERS>
Excluding path due to filter: C:\Program Files (x86)\ => <TRIMMED DEFAULT FILTERS>
Excluding path due to filter: C:\ProgramData\ => <TRIMMED DEFAULT FILTERS>
Excluding path due to filter: C:\System Volume Information\ => <TRIMMED DEFAULT FILTERS>
```

Maybe the filters aren't being set correctly when invoked through the GUI? #2976 also reported some problems with filters not being taken into account through the GUI, but I haven't been able to repro that one via the command line either. I'll have to try to repro these through the GUI and see if anything turns up there.

As for concerns about removing the unrooting of these filters, I can see that (most specifically for paths like *\Windows\*, but less so for *\System Volume Information\* or *\Program Files\*). I can abort the pull request, or may keep some things rooted and unroot only the most unlikely to be interesting things (swapfile.sys, pagefile.sys, System Volume Information, etc.).

**Comment by tygill** [comment]
I just tested the default filters through the GUI, and it looks like they aren't being applied for some reason. I created a test folder with two files - NTUser.dat and Regular.txt, and created a backup for that folder with default filters and ran it once through the GUI and once via the exported command line. The exported command line version correctly ignored NTUser.dat, but the GUI version did not ignore it. I'll try to figure out how the GUI version invokes the backup to see if I can find where the parameter is getting lost.
